File: 1570-final-prices-with-a-special-discount-in-a-shop/final-prices-with-a-special-discount-in-a-shop.cpp
class Solution {
public:
    vector<int> finalPrices(vector<int>& prices) {
        stack<int> s;

        for (int i = 0; i < prices.size(); i++)
        {
            while (s.size() > 0 && prices[i] <= prices[s.top()])
            {
                prices[s.top()] -= prices[i];
                s.pop();
            }
            s.push(i);
        }
        return prices;
    }
};

File: 1570-final-prices-with-a-special-discount-in-a-shop/final-prices-with-a-special-discount-in-a-shop.java
class Solution {
    public int[] finalPrices(int[] prices) {
        Stack<Integer> s = new Stack<>();

        for (int i = 0; i < prices.length; i++)
        {
            while (s.size() > 0 && prices[i] <= prices[s.peek()])
            {
                prices[s.peek()] -= prices[i];
                s.pop();
            }
            s.push(i);
        }
        return prices;
    }
}